rclcpp/include/rclcpp/serialization.hpp

```diff
@@ -0,0 +1,159 @@
+   * \param[out] message The deserialized ROS2 message.
+   */
+  void deserialize_message(
+    const SerializedMessage * serialized_message, void * ros_message) const;
+
+  const rosidl_message_type_support_t * type_support_;
+};
+
+/// Default implementation to (de)serialize a message by using rmw_(de)serialize
+template<typename MessageT>
+class Serialization : public SerializationBase
+{
+public:
+  Serialization()
+  : SerializationBase(rosidl_typesupport_cpp::get_message_type_support_handle<MessageT>())
+  {}
+
+  void serialize_message(
+    const MessageT & ros_message, SerializedMessage & serialized_message) const
+  {
+    SerializationBase::serialize_message(
+      reinterpret_cast<const void *>(&ros_message),
+      &serialized_message);
+  }
+
+  void deserialize_message(
+    const SerializedMessage & serialized_message, MessageT & ros_message) const
+  {
+    SerializationBase::deserialize_message(
+      &serialized_message,
+      reinterpret_cast<void *>(&ros_message));
+  }
+};
+
+template<>
+class Serialization<SerializedMessage>: public SerializationBase
+{
+public:
+  Serialization()
+  : SerializationBase(nullptr)
+  {}
+
+  void serialize_message(
+    const SerializedMessage & ros_message,
+    SerializedMessage & serialized_message) const
+  {
+    (void)ros_message;
+    (void)serialized_message;
+    throw std::runtime_error(
+            "Serialization of serialized message to serialized message is not possible.");
+  }
+
+  void deserialize_message(
+    const SerializedMessage & serialized_message,
+    SerializedMessage & ros_message) const
+  {
+    (void)ros_message;
+    (void)serialized_message;
+    throw std::runtime_error(
+            "Deserialization of serialized message to serialized message is not possible.");
+  }
+};
+
+template<>
+class Serialization<rcl_serialized_message_t>: public SerializationBase
+{
+public:
+  Serialization()
+  : SerializationBase(nullptr)
+  {}
+
+  void serialize_message(
+    rcl_serialized_message_t & ros_message,
+    const SerializedMessage & serialized_message) const
+  {
+    (void)ros_message;
+    (void)serialized_message;
+    throw std::runtime_error(
+            "Serialization of serialized message to serialized message is not possible.");
+  }
+
+  void deserialize_message(
+    const SerializedMessage & serialized_message,
+    rcl_serialized_message_t & ros_message) const
+  {
+    (void)ros_message;
+    (void)serialized_message;
+    throw std::runtime_error(
+            "Deserialization of serialized message to serialized message is not possible.");
+  }
+};
+
+// trait to check if type is the object oriented serialized message
+template<typename T>
+struct is_serialized_message_class : std::false_type
+{};
+
+template<>
+struct is_serialized_message_class<rcl_serialized_message_t>: std::false_type
+{};
+
+template<>
+struct is_serialized_message_class<SerializedMessage>
+  : std::true_type
+{};
```

Comment:
nit: I usually put the traits into their own file, but in this case I think it's okay to have them here. Maybe put them to the top of this file in their own namespace, (i.e. `rclcpp::serialization_traits::is_serialized_message_class`?)